### PR TITLE
[extractor/rtbf] Update stream extractor

### LINKED ISSUE #4656: [RTBF.be] Extration problem for the "Live" streams
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.08.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Belgique

### Provide a description that is worded well enough to be understood

Extration problem for the "Live" streams found here: https://www.rtbf.be/auvio/direct#/

It seems to be blocked because of Subtitle...

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
$ yt-dlp -vU -g --cookies-from-browser firefox https://www.rtbf.be/auvio/direct_le-beau-velo-de-ravel?lid=408218#/
[debug] Command-line config: ['-vU', '-g', '--cookies-from-browser', 'firefox', 'https://www.rtbf.be/auvio/direct_le-beau-velo-de-ravel?lid=408218#/']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.08.08 [3157158] (pip)
[debug] Python 3.9.13 (CPython 64bit) - macOS-10.15.7-x86_64-i386-64bit 
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.1 (setts), ffprobe 5.1
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "/Users/xxx/Library/Application Support/Firefox/Profiles/xc8x3893.default-release/cookies.sqlite"
[Cookies] Extracted 888 cookies from firefox
[debug] Proxy map: {}
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.08.08, Current version: 2022.08.08
yt-dlp is up to date (2022.08.08)
[debug] [RTBF] Extracting URL: https://www.rtbf.be/auvio/direct_le-beau-velo-de-ravel?lid=408218#/
[RTBF] 408218: Downloading webpage
ERROR: 408218: An extractor error has occurred. (caused by KeyError('subtitle')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "/usr/local/Cellar/yt-dlp/2022.8.8/libexec/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 666, in extract
    ie_result = self._real_extract(url)
  File "/usr/local/Cellar/yt-dlp/2022.8.8/libexec/lib/python3.9/site-packages/yt_dlp/extractor/redbee.py", line 283, in _real_extract
    title = data['subtitle']
KeyError: 'subtitle'
```

### PR DESCRIPTION
**Pull request by elyse0**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

In #4479 I changed the `title` extractor from `title` to `subtitle` because, for videos at least, the `subtitle` is more descriptive. https://github.com/yt-dlp/yt-dlp/pull/4479#issuecomment-1200308965

But unfortunately, livestreams don't contain a `subtitle` so we fail extracting it. This is fixed by adding a fallback to `title` in case `subtitle` doesn't exist.

**Livestreams using RedBee**
Also all streams have started using RedBee, but the previous api still returns m3u8 and mpd manifests that contain a 20s video that is completely unrelated, so we need to ignore them since they're 'fake'. https://github.com/yt-dlp/yt-dlp/issues/4656#issuecomment-1214461092

**RedBee DRM formats**
And last but not least, some RedBee formats have drm but we were not detecting it; these were downloaded completely fine but were unplayable.

Fixes #4656

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/redbee.py [hunk 1]
```diff
@@ -69,6 +69,9 @@ def _get_formats_and_subtitles(self, asset_id, **kwargs):
                 fmts, subs = self._extract_m3u8_formats_and_subtitles(
                     format['mediaLocator'], asset_id, fatal=False)
 
+            for index in range(len(fmts)):
+                fmts[index]['has_drm'] = format.get('drm') is not None
```
Comment (pukkandan) [inline yt_dlp/extractor/redbee.py:73]: _extract_m3u8_formats_and_subtitles etc has it's own drm detection. Don't override it

```suggestion
            if format.get('drm'):
                for f in fmts:
                    f['has_drm'] = True
```

File: yt_dlp/extractor/redbee.py [hunk 2]
```diff
@@ -269,8 +272,17 @@ def _real_extract(self, url):
...
             media_id, query={'id': media_id})
-        data = self._parse_json(self._html_search_regex(
-            r'data-media="([^"]+)"', embed_page, 'media data'), media_id)
+
+        media_data = self._html_search_regex(r'data-media="([^"]+)"', embed_page, 'media data', fatal=False)
+        if not media_data:
+            if re.findall(r'<div[^>]+id="[^"]*js-error-expired[^"]*"[^>]+class="(?![^"]*hidden)', embed_page):
+                raise ExtractorError('Livestream has ended.', expected=True)
+            if re.findall(r'<div[^>]+id="[^"]*js-sso-connect[^"]*"[^>]+class="(?![^"]*hidden)', embed_page):
+                self.raise_login_required()
```
Comment (pukkandan) [inline yt_dlp/extractor/redbee.py:281]:
```suggestion
            if re.search(r'<div[^>]+id="[^"]*\bjs-error-expired\b', embed_page):
                raise ExtractorError('Livestream has ended', expected=True)
            if re.search(r'<div[^>]+id="[^"]*\bs-sso-connect\b', embed_page):
                self.raise_login_required()
```
is this insufficient?

File: yt_dlp/extractor/redbee.py [hunk 3]
```diff
@@ -269,8 +272,17 @@ def _real_extract(self, url):
...
+
+            raise ExtractorError('Could not find media data')
+
+        data = self._parse_json(media_data, media_id, fatal=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/redbee.py:285]: why fatal=False?

File: yt_dlp/extractor/redbee.py [hunk 4]
```diff
@@ -280,69 +292,79 @@ def _real_extract(self, url):
         if provider in self._PROVIDERS:
             return self.url_result(data['url'], self._PROVIDERS[provider])
 
-        title = data['subtitle']
+        # In general, we prefer the 'subtitle' because it's a lot more specific than the 'title',
+        # but we need to fallback to 'title' in case 'subtitle' doesn't exist, like for livestreams.
+        # https://github.com/yt-dlp/yt-dlp/issues/4656
+        title = data.get('subtitle') or data['title']
```
Comment (pukkandan) [inline yt_dlp/extractor/redbee.py:298]:
```suggestion
        title = traverse_obj(data, 'subtitle', 'title')
```

File: yt_dlp/extractor/redbee.py [hunk 5]
```diff
@@ -280,69 +292,79 @@ def _real_extract(self, url):
...
-            sources = data.get('sources') or {}
-            for key, format_id in self._QUALITIES:
-                format_url = sources.get(key)
-                if not format_url:
-                    continue
-                height = int_or_none(self._search_regex(
-                    height_re, format_url, 'height', default=None))
+        # The old api still returns m3u8 and mpd manifest for livestreams, but these are 'fake'
+        # since all they contain is a 20s video that is completely unrelated.
+        # https://github.com/yt-dlp/yt-dlp/issues/4656#issuecomment-1214461092
```
Comment (pukkandan) [inline yt_dlp/extractor/redbee.py:306]: Do
```py
m3u8_url = None if data.get('isLive') else traverse_obj(data, 'urlHlsAes128', 'urlHls')
```

This allows you to de-indent this whole large block to it's previous state

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/redbee.py (hunk 3 above)
Comment (elyse0) [reply yt_dlp/extractor/redbee.py:286]: Forgot to remove it x-x  thanks!

File: yt_dlp/extractor/redbee.py (hunk 2 above)
Comment (elyse0) [reply yt_dlp/extractor/redbee.py:282]: Nop, all errors are already inside the page and they remove the 'hidden' class to show it

```html
<div id="js-error-geoloc" class="rtbf-embed-message hidden">
        <figure class="rtbf-embed-message__placeholder">
            <img src="[https://ds1.static.rtbf.be/media/program/image/ng_d41a344a2252c5891b77-324x183@2x.jpg](view-source:https://ds1.static.rtbf.be/media/program/image/ng_d41a344a2252c5891b77-324x183@2x.jpg)" alt="Les minuscules" />
            <div class="www-full-stretch www-pattern"></div>
            <figcaption class="rtbf-embed-message__body">
                <h1>Les minuscules</h1>
                <h2></h2>
                <p>
                    Ce contenu est visible uniquement en Belgique<br>
                    En Belgique ?  Consultez notre
                    <a class="rtbf-embed-message__link" href="[https://www.rtbf.be/auvio/faq](view-source:https://www.rtbf.be/auvio/faq)" target="_top">FAQ</a>
                </p>
                ....
```

Comment (pukkandan) [reply yt_dlp/extractor/redbee.py:282]: gotcha. 
* replace `re.findall` with `re.search`
* Can you explain `id="[^"]*js-error-expired[^"]*"`? Shouldn't it be checking just `id="js-error-expired"`?

Comment (elyse0) [reply yt_dlp/extractor/redbee.py:282]: I was thinking that the regex had to be flexible, I'll change it n-n

**Comment by elyse0** [comment]
@Ottaviocr  I can't reproduce it 🤔  If the problem persists, could you please open an issue? 

```bash
[debug] Command-line config: ['https://www.rtbf.be/auvio/detail_jt-13h?id=2929310', '--cookies-from-browser', 'firefox', '--verbose', '-R', 'infinite', '--fragment-retries', 'infinite', '--restrict-filenames', '--progress', '--downloader=aria2c', '--postprocessor-args', 'thread:1']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.08.08 [3157158f7] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: dbb7be0c3
[debug] Python 3.8.10 (CPython 64bit) - Linux-5.4.0-124-generic-x86_64-with-glibc2.29 (glibc 2.31)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg n5.0.1-8-g11eff7739a-20220712 (setts), ffprobe n5.0.1-8-g11eff7739a-20220712, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "/home/amish/.mozilla/firefox/ckczlffq.default-release/cookies.sqlite"
[Cookies] Extracted 2541 cookies from firefox
[debug] Proxy map: {}
[debug] [RTBF] Extracting URL: https://www.rtbf.be/auvio/detail_jt-13h?id=2929310
[RTBF] 2929310: Downloading webpage
[RTBF] 2929310: Downloading JSON metadata
[RTBF] 2929310: Downloading JSON metadata
[RTBF] 2929310: Downloading JSON metadata
[RTBF] 2929310: Downloading MPD manifest
[RTBF] 2929310: Downloading m3u8 information
[RTBF] 2929310: Downloading ISM manifest
[debug] Sort order given by extractor: res, proto
[debug] Formats sorted by: hasvid, ie_pref, res, proto, lang, quality, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] 2929310: Downloading 1 format(s): 5642+130
[debug] Invoking hlsnative downloader on "https://C9851Ec-az-westeurope.fsly.cdn.ebsd.ericsson.net/rtbf/auvio/assets/2929310_6BA97Bb/materials/4T10JjYJ4K_6BA97Bb/vod-idx.ism/vod-idx-video=5200000.m3u8"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 305
[download] Destination: 16.08.2022-[2929310].f5642.mp4
[download]   0.4% of ~1.09GiB at    1.45MiB/s ETA 07:35 (frag 1/305)
```

**Pull request merged**